src/Orientations.cpp

```diff
@@ -0,0 +1,128 @@
+#include "Orientations.h"
+
+#include "InterpUtils.h"
+
+namespace ale {
+
+  Orientations::Orientations(
+    const std::vector<Rotation> &rotations,
+    const std::vector<double> &times,
+    const std::vector<ale::Vec3d> &avs
+  ) :
+    m_rotations(rotations), m_avs(avs), m_times(times) {
+    if (m_rotations.size() < 2 || m_times.size() < 2) {
+      throw std::invalid_argument("There must be at least two rotations and times.");
+    }
+    if (m_rotations.size() != m_times.size()) {
+      throw std::invalid_argument("The number of rotations and times must be the same.");
+    }
+    if ( !m_avs.empty() && (m_avs.size() != m_times.size()) ) {
+      throw std::invalid_argument("The number of angular velocities and times must be the same.");
+    }
+  }
+
+
+  std::vector<Rotation> Orientations::rotations() const {
+    return m_rotations;
+  }
+
+
+  std::vector<ale::Vec3d> Orientations::angularVelocities() const {
+    return m_avs;
+  }
+
+
+  std::vector<double> Orientations::times() const {
+    return m_times;
+  }
+
+
+  Rotation Orientations::interpolate(
+    double time,
+    RotationInterpolation interpType
+  ) const {
+    int interpIndex = interpolationIndex(m_times, time);
+    double t = (time - m_times[interpIndex]) / (m_times[interpIndex + 1] - m_times[interpIndex]);
+    return m_rotations[interpIndex].interpolate(m_rotations[interpIndex + 1], t, interpType);
+  }
+
+
+  ale::Vec3d Orientations::interpolateAV(double time) const {
+    int interpIndex = interpolationIndex(m_times, time);
+    double t = (time - m_times[interpIndex]) / (m_times[interpIndex + 1] - m_times[interpIndex]);
+    ale::Vec3d interpAv = ale::Vec3d(linearInterpolate(m_avs[interpIndex], m_avs[interpIndex + 1], t));
+    return interpAv;
+  }
+
+  ale::Vec3d Orientations::rotateAvAt(
+    double time,
+    const ale::Vec3d &av,
+    RotationInterpolation interpType,
+    bool invert
+  ) const {
+    Rotation interpRot = interpolate(time, interpType);
+    return interpRot(av);
+  }
+
+
+  ale::State Orientations::rotateStateAt(
+    double time,
+    const ale::State &state,
+    RotationInterpolation interpType,
+    bool invert
+  ) const {
+    Rotation interpRot = interpolate(time, interpType);
+    ale::Vec3d av;
```

Comment:
This needs to default to 0, 0, 0 so that if there's no angular velocities, then it's just the rotation.